### ISSUE TITLE
Is performance of the clone the same as original DB?

### ISSUE THREAD

**Issue by robertwilczynski** [sanderstad/PSDatabaseClone#26]
Hi,

Not sure this is the best place to ask this. We're looking into using PSDatabaseClone in our CI/CD pipeline. I was wondering if clones are a good fit for performance testing and same performance is expected from the clone as from original DB (assuming clone is hosted on the same kind of hardware as original DB)? Have you done any testing or possibly know upfront about any limitations of using a clone for performance testing or should we do our own tests to figure this out?

Thank you,
Rob.

**Comment by sanderstad** [comment]
Hi Rob,

You can certainly use the clones for performance testing. The thing that makes the clones ideal for that is because you have the same statistics and data. Because of that you get the same query plan (assuming you use the same sql server version and compatibility mode) on your clone instance as you would in production.

The entire point of this solution is to give developers, or anyone else who needs to have production data, to get a consistent data set without the need for a lot of disk space.
One scenario I can think of is when you have two clones of the same image on your instance. Clone 1 does not get changed but Clone 2 i.e. gets a different index for a particular table.
Running the same query on Clone 1 and Clone 2 will give you the differences and you can test the performance with the new changes.

Keep in mind that when you clone the database, your data is only present in the image on the network share, and not in your clone. The clone only contains the delta between the image and the clone.
Because of that, when you run a large query, the query will be retrieved data from the image on the network share and get cached in memory on the instance with the clone.
This doesn't have to be a performance decreasing factor but it all depends on the speed of your network. 
I have not thoroughly tested this but I can imagine that this does have an impact on performance testing when you enable the STATISTICS IO or STATISTICS TIME, although we also have to deal with architectures where we have LUNs on SANs which also go over the network.
This is one of those areas that I have not been able to test thoroughly.

Be mindful of the fact that the clone saves the delta. When you for instance, rebuild all your indexes on your clone, all the data from the indexes will be saved in your clone.
That means that when you have i.e. 100 GB of free space and your indexes are 150 GB, your drive will fill up on your clone instance.

To sum things up:
1. You will get the same statistics and query plans because your data is the same as production
2. It's possible to create multiple clones of the same image to compare changes
3. The network speed could be a performance decreasing factor in certain circumstances
4. Keep in mind that changes will be saved on your clone instance.

**Comment by robertwilczynski** [comment]
Thank you for a detailed response Sander. Looking forward to using this. Thank you for the great work.
